docs/source/literal_types.rst

```diff
@@ -168,6 +161,89 @@ For example, compare and contrast what happens when you try appending these type
+
+For example, when we index into a tuple using some int, the inferred type is
+normally the union of the tuple item types. However, if we want just the type
+corresponding to some particular index, we can use Literal types like so:
+
+.. code-block:: python
+
+    tup = ("foo", 3.4)
+
+    # Indexing with an int literal gives us the exact type for that index
+    reveal_type(tup[0])  # Revealed type is 'str'
+
+    # But what if we want the index to be a variable? Normally mypy won't
+    # know exactly what the index is and so will return a less precise type:
+    int_index = 1
+    reveal_type(tup[int_index])  # Revealed type is 'Union[str, float]'
+
+    # But if we use either Literal types or a Final int, we can gain back
+    # the precision we originally had:
+    lit_index: Literal[1] = 1
+    fin_index: Final = 1
+    reveal_type(tup[lit_index])  # Revealed type is 'str'
+    reveal_type(tup[fin_index])  # Revealed type is 'str'
+
+.. _tagged_unions:
+
+Tagged unions
+*************
+
+When you have a union of types, you can normally discriminate between each type
+in the union by using ``isinstance`` checks. For example, if you had a variable ``x`` of
+type ``Union[int, str]``, you could write some code that runs only if ``x`` is an int
+by doing ``if isinstance(x, int): ...``.
+
+However, it is not always possible or convenient to do this. For example, it is not
+possible to use ``isinstance`` to distinguish between two different TypedDicts since
+at runtime, your variable will simply be just a dict.
+
+Instead, what you can do is *label* or *tag* your TypedDicts with a distinct Literal
+type. Then, you can discriminate between each kind of TypedDict by checking the label:
+
+.. code-block:: python
+
+    # Note: if you are using Python 3.7 or earlier, you will need to import
+    # Literal and TypedDict from the typing_extensions module.
+    from typing import Literal, TypedDict, Union
+
+    class NewJobEvent(TypedDict):
+        tag: Literal["new-job"]
+        job_name: str
+        config_file_path: str
+
+    class CancelJobEvent(TypedDict):
+        tag: Literal["cancel-job"]
+        job_id: int
+
+    Event = Union[NewJobEvent, CancelJobEvent]
+
+    def process_event(event: Union[NewJobEvent, CancelJobEvent]) -> None:
+        # Since we made sure both TypedDicts have a key named 'tag', it's
+        # safe to do 'event["tag"]'. This expression normally has the type
+        # Literal["new-job", "cancel-job"], but the check below will narrow
+        # the type to either Literal["new-job"] or Literal["cancel-job"].
+        #
+        # This in turns narrows the type of 'event' to either NewJobEvent 
+        # or CancelJobEvent.
+        if event["tag"] == "new-job":
+            print(event["job_name"])
+        else:
+            print(event["job_id"])
+
+While this feature is mostly useful when working with TypedDicts, you can also
+use the same technique wih regular objects, tuples, or namedtuples.
```

Comment:
Mention that tags can also be proper enum values (using `is`) and classes (using `isinstance()`), not just strings.